src/VisualStudio/Core/Def/Implementation/Snippets/AbstractSnippetExpansionClient.cs

```diff
@@ -393,7 +451,338 @@ public virtual bool TryInsertExpansion(int startPositionInSubjectBuffer, int end
+                        new XElement(
+                            snippetNamespace + "CodeSnippet",
+                            new XAttribute(snippetNamespace + "Format", "1.0.0"),
+                            new XElement(
+                                snippetNamespace + "Header",
+                                new XElement(
+                                    snippetNamespace + "SnippetTypes",
+                                    new XElement(snippetNamespace + "SnippetType", new XText("Expansion"))),
+                                new XElement(snippetNamespace + "Title", new XText(methodName)),
+                                new XElement(snippetNamespace + "Author", "Microsoft"),
+                                new XElement(snippetNamespace + "Description"),
+                                new XElement(snippetNamespace + "HelpUrl"),
+                                new XElement(snippetNamespace + "Shortcut", methodName)),
+                            new XElement(
+                                snippetNamespace + "Snippet",
+                                new XElement(
+                                    snippetNamespace + "Declarations",
+                                    new XElement(
+                                        snippetNamespace + "Literal",
+                                        new XAttribute(snippetNamespace + "Editable", "true"),
+                                        new XElement(snippetNamespace + "ID", new XText("placeholder")),
+                                        new XElement(snippetNamespace + "ToolTip", new XText("")),
+                                        new XElement(snippetNamespace + "Default", new XText("")),
+                                        new XElement(snippetNamespace + "Function"))),
+                                new XElement(
+                                    snippetNamespace + "Code",
+                                    new XAttribute(snippetNamespace + "Language", "csharp"),
+                                    new XCData(template))))));
+
+                var doc = new DOMDocumentClass();
+                if (doc.loadXML(snippet.ToString(SaveOptions.OmitDuplicateNamespaces)))
+                {
+                    var methodSymbols = symbols.OfType<IMethodSymbol>().ToImmutableArray();
+                    _symbols = methodSymbols;
+                    _symbol = null;
+
+                    if (expansion.InsertSpecificExpansion(doc, textSpan, this, LanguageServiceGuid, pszRelativePath: null, out ExpansionSession) == VSConstants.S_OK)
+                    {
+                        Debug.Assert(_symbols == methodSymbols);
+                        Debug.Assert(_symbol == null);
+
+                        // Trigger signature help after starting the snippet session
+                        var workspace = (VisualStudioWorkspace)SubjectBuffer.CurrentSnapshot.GetOpenDocumentInCurrentContextWithChanges()!.Project.Solution.Workspace;
```

Comment:
tagging @jasonmalinowski .  i'm guessing this should be safe... but i'm not sure.  for example, what if we're in the miscworkspace?

Reply:
also, doc the general intution for what's going on here, it's not apparent to me what you're trying to do.

Reply:
⏱️ Added TODO in source which is linked to #50036. I do not know why this works but the more obvious API does not.

Reply:
That looks really sketchy. Is this just trying to get the IEditorCommandHandlerServiceFactory? That should be a direct MEF import then rather than fetching it through the workspace like this.

Reply:
Eliminated the use of `VisualStudioWorkspace` here (uses `Package.GetGlobalService` to access `IComponentModel`). Also eliminated the loop with the `Yield`.